File: promise/src/test/java/com/voxeet/testpromise/all/PromiseTestSimpleAllReject.java
package com.voxeet.testpromise.all;

import com.voxeet.promise.Promise;
import com.voxeet.promise.solve.PromiseExec;
import com.voxeet.testpromise.utils.AndroidMockUtil;

import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class PromiseTestSimpleAllReject {

    @Before
    public void setHandler() {
        Promise.setHandler(AndroidMockUtil.mockMainThreadHandler());
    }

    @Test
    public void test() throws InterruptedException {
        final CountDownLatch latch = new CountDownLatch(1);

        final boolean[] catched = {true};
        System.out.println("executing test");

        Promise.all(new Promise<>(solver -> {
                    try {
                        throw new NullPointerException("oops");
                    } catch (Exception e) {
                        solver.reject(e);
                    }
                }), Promise.resolve("called 2")
        ).then((PromiseExec<List<String>, Void>) (result, solver) -> {
            System.out.println(Arrays.toString(result.toArray()));
            catched[0] = false;
            latch.countDown();
        }).error(error -> {
            error.printStackTrace();
            catched[0] = true;
            latch.countDown();
        });

        //6s are enough
        latch.await(6, TimeUnit.SECONDS);
        if (!catched[0]) {
            throw new IllegalStateException("Expected an error...");
        }
    }
}

File: promise/src/test/kotlin/com/voxeet/testpromise/all/PromiseTestSimpleAllReject.kt
package com.voxeet.testpromise.all

import com.voxeet.promise.HandlerFactory
import com.voxeet.promise.Promise
import com.voxeet.promise.await
import com.voxeet.promise.solve.Solver
import com.voxeet.testpromise.mockedhandler
import com.voxeet.testpromise.registerConfigurationForNoRethrowResolve
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import org.junit.Before
import org.junit.Test
import org.junit.jupiter.api.Assertions.fail
import java.util.*

@OptIn(ExperimentalCoroutinesApi::class)
class PromiseTestSimpleAllReject {
    @Before
    fun setHandler() {
        HandlerFactory.setHandler(mockedhandler())
        registerConfigurationForNoRethrowResolve()
    }

    @Test
    fun test() = runTest {
        val catched = booleanArrayOf(true)

        try {
            Promise.all(
                Promise { solver: Solver<String?> ->
                    try {
                        throw NullPointerException("oops")
                    } catch (e: Exception) {
                        solver.reject(e)
                    }
                }, Promise.resolve("called 2")
            ).then { result: List<String?> ->
                println(result.toTypedArray().contentToString())
                catched[0] = false
                true
            }.await()

            fail("expected error")
        } catch (e: NullPointerException) {
            //expected
        }
    }
}
